runtime/vm/jvminit.c

```diff
@@ -6349,6 +6349,40 @@ protectedInitializeJavaVM(J9PortLibrary* portLibrary, void * userData)
 	Xj9BreakPoint("jvminit");
 #endif
 
+        {
+		IDATA enabled = FIND_AND_CONSUME_ARG(EXACT_MATCH, VMOPT_XXPRINTFLAGSFINALENABLE, NULL);
+		IDATA disabled = FIND_AND_CONSUME_ARG(EXACT_MATCH, VMOPT_XXPRINTFLAGSFINALDISABLE, NULL);
+		if (enabled > disabled) {
+			UDATA maxHeapSize, maxDirectMemorySize;
+			char *howset;
+
+			/*
+			 * Emulate Hotspot -XX:+PrintFlagsFinal output for two specific flags:
+			 *
+			 * [Global flags]
+			 *    size_t MaxHeapSize                              = 4294967296                                {product} {ergonomic}
+			 *  uint64_t MaxDirectMemorySize                      = 3758096384                                {product} {default}
+			 */
+
+			j9tty_printf(PORTLIB, "[Global flags]\n");
+
+			maxHeapSize = (UDATA) vm->memoryManagerFunctions->j9gc_get_maximum_heap_size(vm);
+			howset="ergonomic";
+			if ((findArgInVMArgs( PORTLIB, vm->vmArgsArray, STARTSWITH_MATCH, VMOPT_XMX, NULL, 0) >= 0) ||
+				(findArgInVMArgs( PORTLIB, vm->vmArgsArray, STARTSWITH_MATCH, MAPOPT_XXMAXHEAPSIZE_EQUALS, NULL, 0) >= 0)) {
+				howset = "commandline";
+			}
+			j9tty_printf(PORTLIB, "   size_t MaxHeapSize                              = %-41zu {product} {%s}\n", maxHeapSize, howset);
+
+			maxDirectMemorySize = vm->directByteBufferMemoryMax;
+			howset="ergonomic";
+			if ((findArgInVMArgs( PORTLIB, vm->vmArgsArray, STARTSWITH_MATCH, VMOPT_XXMAXDIRECTMEMORYSIZEEQUALS, NULL, 0) >= 0)) {
+				howset = "commandline";
+			}
+			j9tty_printf(PORTLIB, " uint64_t MaxDirectMemorySize                      = %-41llu {product} {%s}\n", maxDirectMemorySize, howset);
```

Comment:
The format for `maxDirectMemorySize` should not be `llu`, but rather `zu` (like `maxHeapSize`).

Reply:
In fact, the types were aimed at replicating precisely the output from Hotspot. That they were different types was not accidental. Nonetheless, OpenJ9 provides both these values as `UDATA` which I presume is equivalent to `size_t`. Given that, I'll update both to `%zu` which I believe is the appropriate format?

Reply:
The format specifiers need to match the types of the arguments. I wasn't commenting on the `-41` prefix of the format, but just the specifier. If they're `size_t` and `uint64_t` as the comment 6364 suggests, the formats were correct (as `zu` and `llu`).

Reply:
yeah, sorry, didn't mean to put the % there which implied I was removing the field width :)